crates/kubelet/src/state.rs

```diff
@@ -0,0 +1,107 @@
+//! Used to define a state machine of Pod states.
+use log::info;
+
+pub mod default;
+#[macro_use]
+pub mod macros;
+
+use crate::pod::Pod;
+use k8s_openapi::api::core::v1::Pod as KubePod;
+use kube::api::{Api, PatchParams};
+use std::sync::Arc;
+
+/// Represents result of state execution and which state to transition to next.
+#[derive(Debug)]
+pub enum Transition<S, E> {
+    /// Advance to next node.
+    Advance(S),
+    /// Transition to error node.
+    Error(E),
+    /// This is a terminal node of the state graph.
+    Complete(anyhow::Result<()>),
+}
+
+#[async_trait::async_trait]
+/// A trait representing a node in the state graph.
+pub trait State<Provider>: Sync + Send + 'static + std::fmt::Debug {
+    /// The next state on success.
+    type Success: State<Provider>;
+    /// The next state on error.
+    type Error: State<Provider>;
+
+    /// Provider supplies method to be executed when in this state.
+    async fn next(
+        self,
+        provider: Arc<Provider>,
+        pod: &Pod,
+    ) -> anyhow::Result<Transition<Self::Success, Self::Error>>;
+
+    /// Provider supplies JSON status patch to apply when entering this state.
+    async fn json_status(
+        &self,
+        provider: Arc<Provider>,
+        pod: &Pod,
+    ) -> anyhow::Result<serde_json::Value>;
+}
+
+#[async_recursion::async_recursion]
+/// Recursively evaluate state machine until a state returns Complete.
```

Comment:
Have you considered refactoring this to iterate through the states until one returns "complete" instead? That would greatly reduce the stack size, and it would allow the removal of the `async-recursion` crate.

```
while let Transition::Advance(t) = transition {
}
Transition::Complete(transition)
```

Reply:
@thomastaylor312 and I spend several days before settling on this solution. We could not get an iterative approach to compile.